File: p187m/repeated_dna_seq.js
/**
 * @param {string} s
 * @return {string[]}
 */
var findRepeatedDnaSequences = function (s) {
  const [seen, repeated] = [new Set(), new Set()];
  for (let i = 0; i + 9 < s.length; ++i) {
    let ten = s.substring(i, i + 10);
    if (seen.has(ten)) repeated.add(ten);
    else seen.add(ten);
  }
  return Array.from(repeated);
};
// TEST
['AAAAAAAAAAA', 'AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT'].forEach((s) => {
  console.log('Repeated DNA Sequence in', s, '->', findRepeatedDnaSequences(s));
});

File: p187m/repeated_dna_seq.py
from typing import List


class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        ans, seen = set(), set()
        for i in range(len(s) - 9):
            ten = s[i : i + 10]
            if ten in seen:
                ans.add(ten)
            else:
                seen.add(ten)
        return list(ans)


# TESTS
for s, expected in [
    ("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT", ["AAAAACCCCC", "CCCCCAAAAA"]),
    ("AAAAAAAAAAAAA", ["AAAAAAAAAA"]),
]:
    sol = Solution()
    actual = sol.findRepeatedDnaSequences(s)
    print("All repeated 10-letter-long sequences in", s, "->", actual)
    assert sorted(actual) == expected
